homeassistant/components/sensor/tof.py

```diff
@@ -0,0 +1,131 @@
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components import rpi_gpio
+from homeassistant.const import CONF_NAME
+from homeassistant.helpers.entity import Entity
+
+REQUIREMENTS = ['VL53L1X2==0.1.5']
+
+DEPENDENCIES = ['rpi_gpio']
+
+_LOGGER = logging.getLogger(__name__)
+
+LENGTH_MILLIMETERS = 'mm'
+
+CONF_I2C_ADDRESS = 'i2c_address'
+CONF_I2C_BUS = 'i2c_bus'
+CONF_XSHUT = 'xshut'
+
+DEFAULT_NAME = 'VL53L1X'
+DEFAULT_I2C_ADDRESS = 0x29
+DEFAULT_I2C_BUS = 1
+DEFAULT_XSHUT = 16
+DEFAULT_RANGE = 2
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME,
+                 default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_I2C_ADDRESS,
+                 default=DEFAULT_I2C_ADDRESS): vol.Coerce(int),
+    vol.Optional(CONF_I2C_BUS,
+                 default=DEFAULT_I2C_BUS): vol.Coerce(int),
+    vol.Optional(CONF_XSHUT,
+                 default=DEFAULT_XSHUT): cv.positive_int,
+})
+
+
+def init_tof(xshut, level):
+    """XSHUT port LOW resets the device."""
+    rpi_gpio.setup_output(xshut)
+    rpi_gpio.write_output(xshut, level)
+
+
+async def async_setup_platform(hass,
+                               config,
+                               async_add_entities,
+                               discovery_info=None):
+    """Reset and initialize the VL53L1X ToF Sensor from STMicroelectronics."""
+    from VL53L1X2 import VL53L1X  # pylint: disable=import-error
+
+    name = config.get(CONF_NAME)
+    bus_number = config.get(CONF_I2C_BUS)
+    i2c_address = config.get(CONF_I2C_ADDRESS)
+    unit = LENGTH_MILLIMETERS
+    xshut = config.get(CONF_XSHUT)
+
+    await hass.async_add_executor_job(
+        init_tof, xshut, 0
+    )
+    await asyncio.sleep(0.01)
+    await hass.async_add_executor_job(
+        init_tof, xshut, 1
+    )
+    await asyncio.sleep(0.01)
+
+    sensor = await hass.async_add_executor_job(
+        partial(VL53L1X, bus_number)
+    )
+
+    dev = [VL53L1XSensor(sensor, name, unit, i2c_address)]
+
+    async_add_entities(dev, True)
+
+
+class VL53L1XSensor(Entity):
+    """Implementation of VL53L1X sensor."""
+
+    def __init__(self, vl53l1x_sensor, name, unit, i2c_address):
+        """Initialize the sensor."""
+        self._name = name
+        self._unit_of_measurement = unit
+        self.vl53l1x_sensor = vl53l1x_sensor
+        self.i2c_address = i2c_address
+        self._state = None
+
+    @property
+    def name(self) -> str:
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self) -> int:
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def unit_of_measurement(self) -> str:
+        """Return the unit of measurement."""
+        return self._unit_of_measurement
+
+    def measure(self):
+        """Get the latest measurement from VL53L1X."""
+        self.vl53l1x_sensor.open()
+        self.vl53l1x_sensor.add_sensor(self.i2c_address, self.i2c_address)
```

Comment:
Is it ok to add the sensor address multiple times?

Reply:
I did check NagyAttila's driver and double checked now.

`add_sensor` is not an issue, the first argument is an identifier.  Please see the below that, since I used `i2c_address` as the identifier, same address multiple times would be fine.

```
    def add_sensor(self, sensor_id, address):
        self._dev_list[sensor_id] = _TOF_LIBRARY.copy_dev(self._default_dev)
        _TOF_LIBRARY.init_dev(self._dev_list[sensor_id], c_uint8(address))
```
But I guess there is an issue calling `open()` multiple times, as you can see:

```
    def open(self):
        self._i2c.open(bus=self._i2c_bus)
        self._configure_i2c_library_functions()
        self._default_dev = _TOF_LIBRARY.initialise()
        self._dev_list = dict()
```
and `initialise()` would allocate multiple times:
```
VL53L1_Dev_t* initialise()
{
    // Flush stdout and stderr to get the print instantly
    setbuf(stdout, NULL);
    setbuf(stderr, NULL);
    VL53L1_Dev_t *dev = allocate_VL53L1_Dev_t();
...

VL53L1_Dev_t* allocate_VL53L1_Dev_t(void) {
    VL53L1_Dev_t* dev = (VL53L1_Dev_t *) calloc(1, sizeof(VL53L1_Dev_t));
    return dev;
}
```
Isn't there a way to return `open()` to `__init__()` as it was before?

Reply:
We can move it to `async_added_to_hass` coroutine method.

Reply:
GREAT! Is there a code template or any docs about this?

Reply:
I recommend searching the source code. There are plenty of examples.

Reply:
I suppose `async_added_to_hass` is called before `__init__(self, vl53l1x_sensor, name, unit, i2c_address)`, right?
But I need to know about `vl53l1x_sensor` in order to call `vl53l1x_sensor.open()` 
Any hints?
```
    def __init__(self, vl53l1x_sensor, name, unit, i2c_address):
        """Initialize the sensor."""
        self._name = name
        self._unit_of_measurement = unit
        self.vl53l1x_sensor = vl53l1x_sensor
        self.i2c_address = i2c_address
        self._state = None

    def __init(i2c_address):
        self.vl53l1x_sensor.open()
        # self.vl53l1x_sensor.add_sensor(DEFAULT_I2C_ADDRESS, DEFAULT_I2C_ADDRESS)
        # self.vl53l1x_sensor.start_ranging(DEFAULT_I2C_ADDRESS, DEFAULT_RANGE)

    async def async_added_to_hass(self):
        """Handle entity about to be added to hass event."""
        await super().async_added_to_hass()
        last_state = await self.async_get_last_state()
        if last_state:
            await self.hass.async_add_executor_job(
                self.__init(self.i2c_address)
            )
```

Reply:
No, `async_added_to_hass` is called after init after the entity is added to home assistant just before the first state update.

Reply:
Better! I tried this:
```
    def __init__(self, vl53l1x_sensor, name, unit, i2c_address):
        """Initialize the sensor."""
        self._name = name
        self._unit_of_measurement = unit
        self.vl53l1x_sensor = vl53l1x_sensor
        self.i2c_address = i2c_address
        self._state = None

    def __init(i2c_address):
        self.vl53l1x_sensor.open()
        self.vl53l1x_sensor.add_sensor(self.i2c_address, self.i2c_address)

    async def async_added_to_hass(self):
        """Handle entity about to be added to hass event."""
        # await super().async_added_to_hass()
        # last_state = await self.async_get_last_state()
        # if last_state:
        #    await self.hass.async_add_executor_job(
        #        self.__init(self.i2c_address)
        #    )
        await self.hass.async_add_executor_job(
            self.__init, self.i2c_address
        )
```
Initialization passed ok but there's an error at async_update():
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/homeassistant/helpers/entity_platform.py", line 248, in _async_add_entity
    await entity.async_device_update(warning=False)
  File "/usr/local/lib/python3.7/site-packages/homeassistant/helpers/entity.py", line 347, in async_device_update
    await self.async_update()
  File "/config/custom_components/sensor/tof.py", line 146, in async_update
    self.measure
  File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/sensor/tof.py", line 138, in measure
    self.vl53l1x_sensor.start_ranging(self.i2c_address, DEFAULT_RANGE)
  File "/config/deps/lib/python3.7/site-packages/VL53L1X2.py", line 149, in start_ranging
    dev = self._dev_list[sensor_id]
TypeError: 'NoneType' object is not subscriptable
```

Reply:
fyi

```
    def measure(self):
        """Get the latest measurement from VL53L1X."""
        # self.vl53l1x_sensor.open()
        # self.vl53l1x_sensor.add_sensor(self.i2c_address, self.i2c_address)
        self.vl53l1x_sensor.start_ranging(self.i2c_address, DEFAULT_RANGE)
        self.vl53l1x_sensor.update(DEFAULT_I2C_ADDRESS)
        self.vl53l1x_sensor.stop_ranging(self.i2c_address)

    async def async_update(self):
        """Get the latest measurement and update state."""

        await self.hass.async_add_executor_job(
            self.measure
        )
```

Reply:
If `add_entities` is called with second argument `True`, `update` will be called during entity addition, and then the first `update` is called before `async_added_to_hass`. That's why the error is happening I believe.

Reply:
You could do the initialization in `update` if you add some control flow there, eg:
```py
def __init__(self, ...):
    self.init = True
...
def update(self):
    if self.init:
        self.vl53l1x_sensor.open()
        self.vl53l1x_sensor.add_sensor(self.i2c_address, self.i2c_address)
        self.init = False
```

Reply:
"update will be called during entity addition"  Oh! I didn´t know about this parameter.

Yes, this is a nice solution in `update()`! I will remember that! Actually `open()` should be called once and `add_sensor()` for each device in multiple sensors environment. I believe there is no need for this control flow, if `vl53l1x_sensor.add_sensor()` can be called from `__init__`. Please comment.